lib/open-amp/addr_translation.c

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright 2025 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/init.h>
+#include <zephyr/logging/log.h>
+#include <addr_translation.h>
+
+#ifdef CONFIG_OPENAMP_VENDOR_ADDR_TRANSLATION_FILE
+/*
+ * In this file addr_map needs to be defined,
+ * see example in nxp_addr_translation.h
+ */
+#include CONFIG_OPENAMP_VENDOR_ADDR_TRANSLATION_FILE
+#else
+static inline struct regions *get_addr_map(void)
+{
+	return NULL;
+}
+#endif
+
+static struct regions *addr_map;
+
+/**
+ * @brief Translates an offset within an I/O region to a physical address.
+ *
+ * This function first attempts to translate the offset using the driver's
+ * physical address map. If no valid mapping is found, it falls back to the
+ * device physical address map.
+ *
+ * @param io Pointer to the I/O region.
+ * @param offset Offset within the I/O region.
+ *
+ * @return physical address if valid, otherwise METAL_BAD_PHYS.
+ */
+static metal_phys_addr_t translate_offset_to_phys(struct metal_io_region *io,
+						  unsigned long offset)
+{
+	(void)io;
+	size_t i;
+	metal_phys_addr_t tmp_phys;
+	size_t tmp_size;
+	unsigned long tmp_offset = 0;
+
+	if (!addr_map) {
+		return METAL_BAD_PHYS;
+	}
+
+	for (i = 0; i < addr_map->no_pages; i++) {
+		tmp_phys = addr_map->pages[i].phys_start;
+		tmp_size = addr_map->pages[i].size;
+		if ((tmp_offset <= offset) && (offset < tmp_offset + tmp_size)) {
+			return tmp_phys + (offset - tmp_offset);
+		}
+		tmp_offset += tmp_size;
```

Comment:
The offset is  related to the bas address of the io_region. Seems to me not good, but perhaps I'm wrong?
I would expect to see here the use of  `metal_io_phys()` to get the physical address and then check that this address is in the memory translation region.

Reply:
I've pushed a new version.

I renamed and clarified the structures used.
> you have only one io regions
> the addr_map->regions[0].addr must matche with to the base address of your io region

I added the selection of the table of base physical addresses for the I/O region, based on the given physical base address - see `get_phys_map(metal_phys_addr_t phys)`

> the addr_map->regions[] should be continuous.

The physical pages must be mapped continuously in the virtual memory space. If the virtual and physical addresses are the same, then, of course, the pages must be continuous. But, I don't see why I should enforce anything here.

> All this could be avoid if you test io->physmap[0] + offset or 'io->virt + offset'

I validated the offset with the io->size.

Reply:
Ok, I still not convince that this code is robust but it answer for your need and can be fixed later if needed.